ovsdb/client.go

```diff
@@ -105,6 +169,70 @@ func (c *Client) rpc(method string, out interface{}, args ...interface{}) error
 	return nil
 }
 
+// listen starts an RPC receive loop that can return RPC results to
+// clients via a callback.
+func (c *Client) listen() {
+	for {
+		res, err := c.c.Receive()
+		if err != nil {
+			// EOF or closed connection means time to stop serving.
+			if err == io.EOF || strings.Contains(err.Error(), "use of closed network") {
+				return
+			}
+
+			// For any other connection errors, just keep trying.
+			continue
+		}
+
+		// TODO(mdlayher): deal with RPC notifications.
+
+		// Handle any JSON-RPC top-level errors.
+		if err := res.Err(); err != nil {
+			c.doCallback(*res.ID, rpcResponse{
+				Error: err,
+			})
+			continue
+		}
+
+		// Return RPC results via callback.
+		c.doCallback(*res.ID, rpcResponse{
+			Result: res.Result,
+		})
+	}
+}
+
+// addCallback registers a callback for an RPC response for the specified ID,
+// and accepts a channel to return the results on.
+func (c *Client) addCallback(id int, ch chan rpcResponse) {
+	c.cbMu.Lock()
+	defer c.cbMu.Unlock()
+
+	if _, ok := c.callbacks[id]; ok {
+		// This ID was already registered.
+		panicf("OVSDB callback with ID %d already registered", id)
+	}
+
+	c.callbacks[id] = ch
+}
+
+// doCallback performs a callback for an RPC response and clears the
+// callback on completion.
+func (c *Client) doCallback(id int, res rpcResponse) {
+	c.cbMu.Lock()
```

Comment:
`RLock()`, `RUnlock()`, as we don't modify `c.callback[]` in this context?

Reply:
Oh, mybad, you call `delete(c.callback)`...  Then, can it be the normal `Mutex`, instead of `RWMutex`, or do we have reader somewhere?

Reply:
I think it might eventually make sense to have metrics or something that count the number of callbacks in flight, in which case, we'd want rwmutex anyway.